### ISSUE TITLE
ECS Provider does not offer health check configuration

### ISSUE THREAD

**Issue by bbl232** [containous/traefik#1400]
### What version of Traefik are you using (`traefik version`)?
1.2.0

### What is your environment & configuration (arguments, toml...)?
We're running many instances of Traefik in many ECS clusters...
```toml
accessLogsFile = "/var/log/traefik/access.log"

[web]
address = ":8080"

[ecs]

domain = "$CUSTOMER.$STACK"
Cluster = "$CUSTOMER-$STACK"
Watch = true
RefreshSeconds = 15
ExposedByDefault = true
```

### What did you do?
Searched documentation and code for how to configure the health checks

### What did you expect to see?
A label that I could apply to the containers to configure the health check endpoint and interval

### What did you see instead?
The ecs.tmpl and provider code did not show support for health checks

It seems like this wouldn't be too hard to add, but would be great to support. It would be even better if the health check failure could trigger a re-schedule of the task in ecs... or some sort of configurable hook that would execute on a failed health check.

**Comment by timoreimann** [comment]
I suppose the kind of functionality that #1320 brings, just for ECS, would help you?

**Comment by bbl232** [comment]
Yeah it sounds like it! Would be useful to add the concept of hooks as well, it would be nice to have some way to let ECS (or any other orchestrator) know that a container is misbehaving.

**Comment by timoreimann** [comment]
@bbl232 I'm not familiar with ECS, but the orchestrators I know (Marathon, Kubernetes) allow to specify a health check within the application definition sent to the orchestrator's API so that the orchestrator knows when to reschedule a failing instance.

The kind of health check the above mentioned PR brings is more of a readiness check: It allows to take instances into / out of load rotation as they become (un-)ready; this is the case when the instance is failing but could also cover non-critical cases, like an instance losing one of its dependencies which prevent it from serving requests properly.

Usually the two concepts (when is an application instance healthy and when is it ready to serve requests) are distinguished. How would hooks fit in here? Could you explain how Traefik would potentially leverage them with regards to ECS?

Thanks.

**Comment by bbl232** [comment]
@timoreimann ECS has a health check system, but only when using the AWS ELB or ALB as a load balancer. ECS will re-schedule failed/stopped tasks but that only works in the case of a crash, it would be nice to have a real health check that could actually tell wether the application is healthy or not. It would then be useful to have an unhealthy task simply killed, so that ECS re-schedules it.

With the AWS load balancers there is a health check warm up period where you can tell ECS to not worry about failing health checks for a little while after the application starts (usually 5 minutes) during this time, traffic is not sent to that container. After which time, the load balancers will request an endpoint on the containers at a variable interval with a variable timeout and a configurable desired response code, at a configurable endpoint. This takes care of the "Application is serving traffic" type of health check... If a task fails a configurable number of health checks, it is removed from the pool of upstream servers and killed. ECS will then reschedule the task because of the second type of health check which is just a basic "Is the application running" check that is reported by the ECS agent container running on each instance

**Comment by timoreimann** [comment]
@bbl232 a few follow-up comprehension questions:

1. Would Traefik be responsible for killing an unresponsive task?
1. Would that involve actually terminating the task's process?
1. How would that be possible if Traefik and the task run on different machines?

**Comment by bbl232** [comment]
@timoreimann 

1. My thought is that Traefik would be responsible for restarting the container if "Application is serving" health checks are failing.
2. No, the actual process inside the container would now be stopped, the task itself would be.
3. Traefik could use the ECS StopTask call to the AWS SDK... documentation linked below...

[StopTask](http://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#ECS.StopTask)

ECS will re-schedule tasks (because they are part of a service) automatically after that.

This should probably be configurable via a label as well... notify vs act or something like that.

Another option is to have Traefik populate a CloudWatch metric for each services health check status. That would allow users to take action triggered by CloudWatch metric alarms. That enables complete control over what happens when a health check is failing. It would be possible to provide a CloudWatch metric name in a label that Traefik reports a simple 0 or 1 to, or maybe response time.

Documentation also linked below...

[PutMetricData](https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#CloudWatch.PutMetricData)

What would be REALLY nice is to have both of these options available... it would allow for a simple restart of the service if you enable that via a label with the health check data, or the user can set up much more complex, custom actions based on an alarm in AWS.

What do you think? :)

**Comment by timoreimann** [comment]
hey @bbl232, thanks for the detailed feedback, and sorry for the late response!

From my experience with other orchestrators, it seems a bit unconventional that the load balancer is responsible (and required) to signal the termination of a process/task. But if that's the way it's supposed to be handled in ECS, then I'm okay with proceeding accordingly in Traefik.

It sounds like the StopTask approach might be easier to do, so maybe that would be a good first step. PRs welcome. :-)

**Comment by zyzop** [comment]
Hello, ECS restarts tasks by itself. You configure how much tasks(containers) you want, and if some containers die then ECS starts them until you have a configured amount running.

**Comment by bbl232** [comment]
@zyzop it is true that ECS will restart tasks that die, that is only true if a container has stopped running. If the container is still running but not serving traffic, ECS will happily leave the task running without any indication that the service is unhealthy

**Comment by zyzop** [comment]
@bbl232 yep, that's true, but anyway you don't need traefik to care about this.
Traefik may kill tasks allright, but ECS will start them by itself a bit later.

**Comment by ldez** [comment]
Closed by #2421
